### ISSUE TITLE
The unix_tai time format appears to be off by 8 seconds

### ISSUE THREAD

**Issue by r-owen** [astropy/astropy#11245]
### Description

The zero point for the new time format `unix_tai` appears to be off by 8 seconds.

### Expected behavior

The new time format `unix_tai` is defined to match the `CLOCK_TAI` Linux real-time clock (and I was looking forward to using it, for that reason).

For a current date  I expect CLOCK_TAI  - CLOCK_UTC = 37 seconds and that is what I observe on a linux system whose leap second table is properly maintained by ntp.
When I try the same computation using Astropy I get 29 seconds.

### Actual behavior

Astropy.time seems to believe that CLOCK_TAI - CLOCK_UTC is 29 seconds for current dates. That is what is shown as an example in your docs and is coded in the unit test, so you clearly have a reason for this. But I don't see how it is compatible with CLOCK_TAI.

### Steps to Reproduce

1: Find a linux system that has a current leap second table (linux is the only OS I know of that supports the CLOCK_TAI real time clock).
2: Use a reasonably recent Python 3 (I use 3.7)
3: Use AstroPy 4.1 or the current dev version

```python
import time
import astropy.time
clock_tai = getattr(time, "CLOCK_TAI", 11)
now_utc_unix = time.time()
now_tai_unix = time.clock_gettime(clock_tai)
astropy_now_utc = astropy.time.Time(now_utc_unix, format="unix")
print(now_tai_unix - now_utc_unix)
print(astropy_now_utc.unix_tai - now_utc_unix)
```

### System Details

>>> import platform; print(platform.platform())
Linux-3.10.0-1062.12.1.el7.x86_64-x86_64-with-centos-7.7.1908-Core
>>> import sys; print("Python", sys.version)
Python 3.7.8 | packaged by conda-forge | (default, Jul 23 2020, 03:54:19) 
[GCC 7.5.0]
>>> import numpy; print("Numpy", numpy.__version__)
Numpy 1.19.1
>>> import astropy; print("astropy", astropy.__version__)
astropy 4.3.dev452+gc3132e70a
>>> import scipy; print("Scipy", scipy.__version__)
Scipy 1.5.2
>>> import matplotlib; print("Matplotlib", matplotlib.__version__)
Matplotlib 3.0.3

**Comment by r-owen** [comment]
Here is the output I see:
```
37.00000762939453
29.0
```

**Comment by mhvk** [comment]
The 29 is expected at least according to the documentation - see https://docs.astropy.org/en/latest/api/astropy.time.TimeUnixTai.html#astropy.time.TimeUnixTai

The difference of 8 may be because the zero point is `1970-01-01 00:00:08 TAI`? I.e., the difference in number of seconds should be the number of leap seconds *since 1970*.

p.s. Your code gives a different result on my Debian system for the built-in package (I have python 3.9.1):
```
2.956390380859375e-05
29.0
```

**Comment by r-owen** [comment]
@mhvk the reason you see a different result is that your CLOCK_TAI is not configured. You are seeing the classic issue that CLOCK_TAI is returning the same value as CLOCK_UTC. It takes some care to configure ntp to set the leap second table (we are motivated to go to the trouble because we use TAI to run our telescopes at Vera Rubin Observatory). If you run ptp instead of ntp I think it will always set the leap second table, but thats a much less popular option (one we are considering).

I understand that the 8 second difference probably comes from the (surprising to me) documented zero point. But tai_unix is also documented to match CLOCK_TAI and in my own tests it does not. My own tests show that CLOCK_TAI has the value CLOCK_TAI = CLOCK_UTC + 37 seconds. I have not found a formal specification for CLOCK_TAI, but the 37 second difference is certainly what we expected and what we rely on for our telescope control systems.

**Comment by taldcroft** [comment]
@r-owen - the definition currently in astropy is that both `unix` and `unix_tai` count seconds since the standard `unix` Epoch time of `1970-01-01 00:00:00 UTC`, but `unix_tai` includes all seconds whereas `unix` does not include leap seconds.  There have been 29 leap seconds since that Epoch time, hence the difference of 29.
```
In [2]: t = Time('1970-01-01 00:00:00', scale='utc')

In [3]: t.unix
Out[3]: 0.0

In [4]: t.unix_tai
Out[4]: 8.200000198854696e-05
```
See https://github.com/astropy/astropy/issues/10052#issuecomment-605652478 for context about that 82 us.

Now it sounds like you are saying that `unix_tai` should count seconds since `1970-01-01 00:00:00 TAI`?

@r-owen - If this is the case, can you figure out what went wrong in the discussion of your #10052 issue and #10081 which you and @timj looked at?

**Comment by taldcroft** [comment]
This document https://www.cl.cam.ac.uk/~mgk25/posix-clocks.html is the only reference I know about, and it is basically ambiguous about the definition. 

**CLOCK_TAI** 
This clock is only available when the system knows International Atomic Time (TAI) with at least an accuracy of 1 s. _The only difference between TAI and UTC is that TAI is never corrected by leap seconds, therefore TAI is a few whole seconds ahead of UTC (one second more after every UTC leap second)_. Some time broadcasting services such as GPS provide both TAI and UTC (e.g., by publishing a scale linked to TAI plus the difference to UTC). TAI is needed for instance to control processes (e.g., astronomical observations, navigation, etc.) where leap seconds are undesirable. CLOCK_TAI is handled very similarly to CLOCK_UTC in that it becomes unavailable when the clock filter algorithm estimates the accuracy of its output to be worse than 1 s.

Basically it doesn't explicitly say if that whole number difference at 1970-01-01 is 0 (makes sense since this is the start of UNIX time) or 8 (total number of leap seconds at that point).

**Comment by taldcroft** [comment]
If this needs to be fixed, then we should talk about strategy. Maybe we just change the epoch time to be 1970-01-01 00:00:00 TAI and backport, advertising this as a breaking change due to a conceptual mistake in the original implementation. The other plan would be introducing yet another `unix_tai` time with that epoch.

**Comment by mhvk** [comment]
In the end, I think we have to match `CLOCK_TAI` however that is actually implemented in current systems. Hopefully, those are consistent...

I fear we need to go the "advertise as a breaking change route" - having two implementations is just going to be super-confusing.

**Comment by taldcroft** [comment]
@r-owen - are you confident that most / all appropriately configured POSIX machines will give this same answer? Is there an easy way to know?

**Comment by r-owen** [comment]
I am reasonably confident because:
* It is true for all our computers running ntp. These include a mix of linux, "compact RIO" and PXI computers, all running their own different flavors of linux.
* I can't imagine what else unix is going to do. It uses the standard leap second table and that has a value of 37 for today's date.

But I can't prove it's true for all linux systems. It be nice to hear from somebody synching time with ptp.

**Comment by r-owen** [comment]
Here is a non-authoritative answer: https://superuser.com/a/1163770 that explains how to configure ntp to maintain the leap second table and which states that the offset is 37 seconds.

**Comment by taldcroft** [comment]
OK, I had entirely forgotten about PR #11143  that got merged only a month again, but it feels like a century ago at this point. I think this does what we agreed we shouldn't do because it would be too confusing. Gahh.

**Comment by taldcroft** [comment]
@r-owen - thanks for the feedback and reference. That's good enough for me. Are you OK using `unix_ptp`? Should we change `unix_tai` still?

**Comment by mhvk** [comment]
Well, I had completely missed that one! But I did look at the unix_tai documentation and did not see it mentioned, so a minimal follow-up would be to add a cross-reference in the documentation of the two formats... (and probably note there that `unix_ptp` is likely what one wants...)

**Comment by r-owen** [comment]
I worry that having both unix_tai and unix_ptp, and having them give different answers, will be very confusing. Especially as unix_tai is documented to match CLOCK_TAI, which it doesn't.

I suggest getting rid of one of them and having the other match CLOCK_TAI (e.g. 37 second difference in 2020-01).

Ignoring the historical situation, I think the better name is `unix_tai` because:
* It is primarily about matching CLOCK_TAI
* Those of us who use CLOCK_TAI often use ntp instead of ptp.

My suggestion is to do this in 2 stages:
1) Change unix_tai to match unix_ptp and document that unix_ptp is deprecated.
2) Remove unix_ptp in some future release.

But if you want less breakage and don't mind a more confusing (to me, at least) name, then just drop unix_tai and refer people to unix_ptp.

**Comment by taldcroft** [comment]
That sounds like a plan. In fact the `unix_ptp` format is not yet released, so I think we can just revert #11143 and move forward with the fix to `unix_tai`.

**Comment by maxnoe** [comment]
The basic question is if the reference for `unix_tai` should be `1970-01-01T00:00:00 TAI` or `1970-01-01T00:00:00 UTC`.

I thought it was intentional that `unix_tai` uses `1970-01-01T00:00:00 UTC` and that it is really a difference to what ptp uses `1970-01-01T00:00:00 TAI`.

If there is no system that uses `1970-01-01T00:00:00 UTC` as reference for counting TAI seconds, the format makes no sense and should probably even receive the fix in a bugfix release to 4.2 and having a separate `unix_ptp` is not necessary.


However, when investigating this, I could not configure my system to actually do this, so I had no way to check.

I used this C-Code:
```C
#include "time.h"
#include "stdio.h"



int main() {
	struct timespec utc;
	struct timespec tai;

	clock_gettime(CLOCK_REALTIME, &utc);
	clock_gettime(CLOCK_TAI, &tai);

	printf("UTC: %ld\n", utc.tv_sec);
	printf("TAI: %ld\n", tai.tv_sec);

	return 0;
}
```

**Comment by maxnoe** [comment]
If someone could point me to a resource to what I need to configure to have the leap seconds table taken into account for this, I would be glad.

**Comment by r-owen** [comment]
@maxnoe ntpd should auto-update the leap second table. See https://kb.meinbergglobal.com/kb/time_sync/ntp/configuration/ntp_leap_second_file#versions_of_ntpd_behaving_differently

As to the epoch...previous discussion has already gone into this, but I'll say again that:

1) unix_tai is documented to match CLOCK_TAI (if your linux is properly configured; that clock seems to be unique to linux), and that was the motivation for adding it in the first place.

2) If your operating system is properly configured then CLOCK_TAI is off from CLOCK_UTC by the current number of leap seconds (presently 37); if not, the two match and that is unfortunately rather common (if you do a web search you'll see many questions about that). I have not found an official spec for the epoch of CLOCK_TAI, but we have only found instances where the it is as stated. In addition to our systems running ntpd, a colleague sees this on a system running ptp, and that is as expected given the pending unix_ptp format.

**Comment by maxnoe** [comment]
Ok, then I would go for just calling it a bug and fix it (event in 4.2 where it first appeared), making `unix_ptp` unnecessary or just an alias.

**Comment by mhvk** [comment]
Sounds like a plan!

**Comment by maxnoe** [comment]
Ok then, I'll just go ahead and open the PR targeting 4.2.x

**Comment by pllim** [comment]
> open the PR targeting 4.2.x

Not sure what you meant but please don't open the PR directly to 4.2.x branch, but rather open it against `master` and milestone to the next 4.2.x release. Thanks!

**Comment by maxnoe** [comment]
@pllim  I just meant which part of the Changelog I will fill in ;)

I am not a maintainer, so @mhvk or youwill have to set the milestone.

**Comment by maxnoe** [comment]
@r-owen you write use astropy 4.1. I thought `unix_tai` was only introduced in `4.2`, is that wrong?

If yes, we need to milestone 4.1.x, not 4.2.x.

**Comment by pllim** [comment]
Re: 4.1 vs 4.2 -- As per discussions on Slack, there is no more planned 4.1.x release, so this has to go into 4.2.x, but we can put a note in Known Issues to state that this bug exists in 4.1. And then remove this Known Issues entry at 5.0 release (the next LTS). cc @eteq

**Comment by maxnoe** [comment]
Monkey patching this is possible by the way:

```
from astropy import version
from astropy.time import TimeUnixTai

if version.major == 4 and version.minor <= 2 and version.bugfix <= 0:
    # fix for astropy #11245
    TimeUnixTai.epoch_val = '1970-01-01 00:00:00.0'
    TimeUnixTai.epoch_scale = 'tai'

```
